gdal/port/cpl_userfaultfd.cpp

```diff
@@ -0,0 +1,214 @@
+ * Copyright (c) 2018, James McClain <james.mcclain@gmail.com>
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a
+ * copy of this software and associated documentation files (the "Software"),
+ * to deal in the Software without restriction, including without limitation
+ * the rights to use, copy, modify, merge, publish, distribute, sublicense,
+ * and/or sell copies of the Software, and to permit persons to whom the
+ * Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included
+ * in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+ * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+ * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ ****************************************************************************/
+
+#include <cstdlib>
+#include <cstring>
+
+#include <fcntl.h>
+#include <poll.h>
+#include <pthread.h>
+#include <unistd.h>
+
+#include <sys/ioctl.h>
+#include <sys/mman.h>
+#include <sys/stat.h>
+#include <sys/syscall.h>
+#include <sys/types.h>
+#include <linux/userfaultfd.h>
+
+#include "cpl_vsi.h"
+#include "cpl_userfaultfd.h"
+
+
+#define BAD_MMAP (reinterpret_cast<void *>(-1))
+
+
+void * fault_handler(void * ptr);
+void uffd_cleanup(void * ptr);
+
+struct cpl_uffd_context {
+  int uffd;
+  struct uffdio_register uffdio_register;
+  char filename[0x1000];
+  uint64_t page_size;
+  void * page_ptr;
+  uint64_t vma_size;
+  void * vma_ptr;
+};
+
+
+void uffd_cleanup(void * ptr)
+{
+  struct cpl_uffd_context * ctx = static_cast<struct cpl_uffd_context *>(ptr);
+
+  if (!ctx) return;
+
+  if (ctx->uffd != -1) {
+    ioctl(ctx->uffd, UFFDIO_UNREGISTER, &ctx->uffdio_register);
+    close(ctx->uffd);
+  }
+  if (ctx->page_ptr && ctx->page_size)
+    munmap(ctx->page_ptr, ctx->page_size);
+  if (ctx->vma_ptr && ctx->vma_size)
+    munmap(ctx->vma_ptr, ctx->vma_size);
+  free(ctx);
+
+  return;
+}
+
+void * fault_handler(void * ptr)
+{
+  struct cpl_uffd_context * ctx = static_cast<struct cpl_uffd_context *>(ptr);
+  struct uffd_msg msg;
+  struct uffdio_copy uffdio_copy;
+  struct pollfd pollfd;
+
+  pthread_cleanup_push(uffd_cleanup, ctx);
+
+  pollfd.fd = ctx->uffd;
+  pollfd.events = POLLIN;
+
+  // Open asset for reading
+  VSILFILE * file = VSIFOpenL(ctx->filename, "rb");
+
+  if (!file) return nullptr;
+
+  // Loop indefinitely
+  while(1) {
+    uintptr_t fault_addr;
+    uint64_t offset;
+
+    // Poll for event
+    if (poll(&pollfd, 1, 16) == -1) break;
```

Comment:
do we need a timeout ? Why infinite timeout wouldn't work ?

Reply:
so the close() in unhook() causes an error event ?

Reply:
Yes, that is the reasoning.  I am open to other approaches